clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -561,12 +562,21 @@ private static int parseAcks(String acksString) {
      * @throws org.apache.kafka.common.errors.AuthorizationException fatal error indicating that the configured
      *         transactional.id is not authorized. See the exception for more details
      * @throws KafkaException if the producer has encountered a previous fatal error or for any other unexpected error
+     * @throws InterruptException if the thread is interrupted while blocked
      */
     public void initTransactions() {
         throwIfNoTransactionManager();
         TransactionalRequestResult result = transactionManager.initializeTransactions();
         sender.wakeup();
-        result.await();
+        try {
+            if (!result.await(requestTimeoutMs, TimeUnit.MILLISECONDS)) {
+                transactionManager.transitionToFatalError(
+                        new TimeoutException("Timeout expired while initializing the transaction in " + requestTimeoutMs + "ms."));
```

Comment:
This should probably say "Timeout expired while initializing transactional state ..."